contracts/javascore/xcall-multi-protocol/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java

```diff
@@ -0,0 +1,356 @@
+
+import score.Address;
+import score.BranchDB;
+import score.Context;
+import score.DictDB;
+import score.RevertedException;
+import score.UserRevertedException;
+import score.VarDB;
+import score.annotation.EventLog;
+import score.annotation.External;
+import score.annotation.Optional;
+import score.annotation.Payable;
+import scorex.util.HashMap;
+
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+
+public class CallServiceImpl implements CallService, FeeManage {
+    public static final int MAX_DATA_SIZE = 2048;
+    public static final int MAX_ROLLBACK_SIZE = 1024;
+    public static String nid;
+
+    private final VarDB<BigInteger> sn = Context.newVarDB("sn", BigInteger.class);
+    private final VarDB<BigInteger> reqId = Context.newVarDB("reqId", BigInteger.class);
+
+    private final DictDB<BigInteger, CallRequest> requests = Context.newDictDB("requests", CallRequest.class);
+    private final DictDB<BigInteger, CSMessageRequest> proxyReqs = Context.newDictDB("proxyReqs", CSMessageRequest.class);
+    private final BranchDB<byte[], DictDB<String, Boolean>> pendingReqs = Context.newBranchDB("pendingReqs", Boolean.class);
+    private final BranchDB<byte[], DictDB<String, Boolean>> pendingResponses = Context.newBranchDB("pendingResponses", Boolean.class);
+
+    // for fee-related operations
+    private final VarDB<Address> admin = Context.newVarDB("admin", Address.class);
+    private final VarDB<BigInteger> protocolFee = Context.newVarDB("protocolFee", BigInteger.class);
+
+    public CallServiceImpl(String networkId) {
+        nid = networkId;
+    }
+
+    /* Implementation-specific external */
+    @External(readonly=true)
+    public String getNetworkAddress() {
+        return new NetworkAddress(nid, Context.getAddress()).toString();
+    }
+
+    private void checkCallerOrThrow(Address caller, String errMsg) {
+        Context.require(Context.getCaller().equals(caller), errMsg);
+    }
+
+    private void onlyOwner() {
+        checkCallerOrThrow(Context.getOwner(), "OnlyOwner");
+    }
+
+    private BigInteger getNextSn() {
+        BigInteger _sn = this.sn.getOrDefault(BigInteger.ZERO);
+        _sn = _sn.add(BigInteger.ONE);
+        this.sn.set(_sn);
+        return _sn;
+    }
+
+    private BigInteger getNextReqId() {
+        BigInteger _reqId = this.reqId.getOrDefault(BigInteger.ZERO);
+        _reqId = _reqId.add(BigInteger.ONE);
+        this.reqId.set(_reqId);
+        return _reqId;
+    }
+
+    private void cleanupCallRequest(BigInteger sn) {
+        requests.set(sn, null);
+    }
+
+    @Override
+    @Payable
+    @External
+    public BigInteger sendCallMessage(String _to, ConnectionPair[] connections, byte[] _data,  @Optional byte[] _rollback) {
+        Address caller = Context.getCaller();
+        // check if caller is a contract or rollback data is null in case of EOA
+        Context.require(caller.isContract() || _rollback == null, "RollbackNotPossible");
+
```

Comment:
we can move rollback check to left so that it skips contract check if rollback is null